presto-pinot/presto-pinot-lib/src/main/java/com/facebook/presto/pinot/PinotConnectorPlanOptimizer.java

```diff
@@ -0,0 +1,263 @@
+ */
+package com.facebook.presto.pinot;
+
+import com.facebook.presto.expressions.LogicalRowExpressions;
+import com.facebook.presto.pinot.PinotClusterInfoFetcher.TimeBoundary;
+import com.facebook.presto.pinot.query.PinotFilterExpressionConverter;
+import com.facebook.presto.pinot.query.PinotQueryGenerator;
+import com.facebook.presto.pinot.query.PinotQueryGeneratorContext;
+import com.facebook.presto.spi.ColumnHandle;
+import com.facebook.presto.spi.ConnectorPlanOptimizer;
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.ConnectorTableHandle;
+import com.facebook.presto.spi.TableHandle;
+import com.facebook.presto.spi.VariableAllocator;
+import com.facebook.presto.spi.function.FunctionMetadataManager;
+import com.facebook.presto.spi.plan.FilterNode;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.PlanNodeIdAllocator;
+import com.facebook.presto.spi.plan.PlanVisitor;
+import com.facebook.presto.spi.plan.TableScanNode;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.facebook.presto.spi.type.TypeManager;
+import com.google.common.collect.ImmutableList;
+
+import javax.inject.Inject;
+
+import java.util.ArrayList;
+import java.util.IdentityHashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.Objects.requireNonNull;
+
+public class PinotConnectorPlanOptimizer
+        implements ConnectorPlanOptimizer
+{
+    private final PinotQueryGenerator pinotQueryGenerator;
+    private final TypeManager typeManager;
+    private final FunctionMetadataManager functionMetadataManager;
+    private final LogicalRowExpressions logicalRowExpressions;
+    private final PinotConfig pinotConfig;
+    private final PinotConnection pinotConnection;
+
+    @Inject
+    public PinotConnectorPlanOptimizer(PinotQueryGenerator pinotQueryGenerator, TypeManager typeManager, FunctionMetadataManager functionMetadataManager, LogicalRowExpressions logicalRowExpressions, PinotConfig pinotConfig, PinotConnection pinotConnection)
+    {
+        this.pinotQueryGenerator = requireNonNull(pinotQueryGenerator, "pinot query generator is null");
+        this.typeManager = typeManager;
+        this.functionMetadataManager = functionMetadataManager;
+        this.logicalRowExpressions = logicalRowExpressions;
+        this.pinotConfig = pinotConfig;
+        this.pinotConnection = pinotConnection;
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode maxSubplan, ConnectorSession session, VariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, boolean preExchange)
```

Comment:
I guess with #13521, we don't need `preExchange` anymore.

Reply:
Yeah I will remove it .. I had added that since you had said that other connectors might need it :-) Pinot does not make use of it anyway.